From 52b17145f4b646912a33224e24616c3f78316da8 Mon Sep 17 00:00:00 2001
From: Szehon Ho <szehon@cloudera.com>
Date: Mon, 4 May 2015 11:20:36 -0700
Subject: [PATCH 0151/1164] HIVE-10571 : HiveMetaStoreClient should close
 existing thrift connection before its reconnect
 (Chaoyu Tang via Szehon)

---
 .../hadoop/hive/metastore/HiveMetaStoreClient.java |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
index bc9ab4a..638025d 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
@@ -302,6 +302,7 @@ public void reconnect() throws MetaException {
       throw new MetaException("For direct MetaStore DB connections, we don't support retries" +
           " at the client level.");
     } else {
+      close();
       // Swap the first element of the metastoreUris[] with a random element from the rest
       // of the array. Rationale being that this method will generally be called when the default
       // connection has died and the default connection is likely to be the first array element.
@@ -486,7 +487,7 @@ public void close() {
         client.shutdown();
       }
     } catch (TException e) {
-      LOG.error("Unable to shutdown local metastore client", e);
+      LOG.debug("Unable to shutdown metastore client. Will try closing transport directly.", e);
     }
     // Transport would have got closed via client.shutdown(), so we dont need this, but
     // just in case, we make this call.
-- 
1.7.9.5

